Add: temporary map labels

SlySven opened the pull request:
These will last only for the duration of a session and will NOT be saved into a binary or JSON format map file later. As far as possible the creation, modification or deletion of such a label will also NOT cause Mudlets's map "auto-save" functionality to be activated.

This should provide an alternative way to work around and close #6265.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Add: temporary map labels

These will last only for the duration of a session and will NOT be saved
into a binary or JSON format map file later. As far as possible the
creation, modification or deletion of such a label will also NOT cause
Mudlets's map "auto-save" functionality to be activated.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
This is untested (but it does compile) code - which is why I am marking it as a draft at this point. Ironically it took me longer to work out the code to save the binary and JSON Map files than it did to write the rest of the code... :astonished:

SlySven commented:
Discord commentary:
> Vadi — Today at 16:36
> Looks good at a glance! Thank you
> I'm not sure we need to expose that functionality in the UI, would anyone ever need to create a temp label that doesn't save? Might be more clutter than help

...
> SlySven — Today at 19:35
> When I first saw the dialogue I wasn't sure whether it was only used for creating labels or also for editing them (like wot I dun for the "Custom Line Properties") - eventually I deduced that it was only for creating them.
>
>As I noted above the arguments to create a Text map label has now reached a total of 19 - maybe we are going to have to rethink that? In a somewhat related line, there isn't a Lua API to either find out if a label is a temporary one - or to change that option after it has been created - now the option to change it would be awkward because, by definition it is either changing something that does need to be saved in the map file into something that is not, or vice versa and so must activate the map auto-save code. OTOH a getter to discover if a label is temporary or not would not have that complication. So, do we need/want either?

> Vadi — Today at 20:06
> We can return that info as part of https://wiki.mudlet.org/w/Manual:Lua_Functions#getMapLabel ?
>>maybe we are going to have to rethink that?
> 
> We should keep the existing API and also allow passing in a table of key-value arguments - that will help with clarity

In doing some testing I am finding that the "Create Label" UI from the 2D Mapper generates a new map label before it is finished (so as to display/preview what the user is setting in the UI) but that in itself is storing a map label in the area concerned which is now tickling the auto-save map code **before** the user has committed to creating the label whether temporary or not. As such it is not simple to prevent that so I will ripped that part of the code out of the PR - so that the Lua API is the only way to create temporary map labels... :slightly_frowning_face:

SlySven committed:
Revise: tweak the map auto-save code to debug faulty invocations

Also add debugging code to track where map autosaves are being triggered;
this revealed that one place in `(void) T2DMap::mouseMoveEvent(...)` was
bogus and causing an unnecessary auto-save. It also seemed useful to
disable initiating an auto-save whilst the profile was being loaded and all
the scripts were being evaluated (for correctness amongst other reasons!)

As the `#define` needed to control the debugging code needs to be in more
than one file its definition has been placed in the project files for each
build system rather than being left at the top of a `.h` or `.cpp` file.
This lead me to promulgate #6314 to do the same for all the other bits of
debugging code that has (mainly, if not entirely by myself) been placed
around the projects source code.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: remove temporary map label creation capability from 2D mapper UI

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 committed:
Merge branch 'development' into Add_temporaryMapLabels

demonnic commented:
the Lua API is the only way to create temp everything else, so this should be fine

SlySven commented:
I have an improvement to follow - but it ain't on my current PC/OS and my local (LAN) repository PC which I use to transfer/store all my changes locally between machines is refusing to start because the ancient battery has finally expired after being powered down for a month - I hope to post it up in the next 24 hours!

SlySven committed:
Revise: enable debugging code in Qmake as well as the CMake project file!

Forgot to do this - will need to disable the code when converting this PR
from *draft* to *ready* status.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented on src/TMap.cpp:
```diff
@@ -2217,8 +2229,17 @@ void TMap::deleteMapLabel(int area, int labelId)
+        // one that the QMap<T1, T2>::take(const T1&) has created for us in the
+        // absence of an actual TMapLabel.
+        // The TMapLabel default constructor sets the temporary member to false
+        // so we can safely rely on it being true for a permanent one:
```
:blush: Oops, got that the wrong way around! Need to change that to "... true for a temporary one:"

SlySven replied:
Fixed by: https://github.com/Mudlet/Mudlet/pull/6285/commits/18e64d1b843c0144fbd752f47a8215e3954c9ce4

SlySven committed:
Merge branch 'development' into Add_temporaryMapLabels

To try to fix failing CI/CB processes...!

Signed-off by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: fix CMake builds

I was adding `#define`s before the `mudlet` executable target had been
defined!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: correct a comment error

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Add_temporaryMapLabels

SlySven commented:
This is ready for review but not a merge as there is some debugging code to be disabled/removed before merging should be done.

Delwing commented:
Looks ok, to be honest would be way more clear and easy to review if `setUnsaved` was in separate PR.

SlySven commented:
> Looks ok, to be honest would be way more clear and easy to review if `setUnsaved` was in separate PR.

Trouble is, I found I needed it to track down why saves were being initiated whilst working with the temporary labels of this PR. The `mpMap->mUnsavedMap = true;` on line 4476 in `(void) T2DMap::mouseMoveEvent(QMouseEvent* event)` was bogus. 

*Then* I found that the MUD specific packages I was testing with was also causing an unneeded triggering as well (it was initiating a set of map colours unconditionally when they were already present in the saved map file - instead the code ought to have only assign a particular colour to an ID if it didn't exist or was different!)

TBH I was planning to leave the code in place but disabled/excluded from compilation but there are a couple of other options i suppose:
1). After determining that the additional temporary map labels work (and do not cause a map autosave to be initiated when they are created/modified/removed - but permanent ones DO) remove/revert the code that tracks the map autosave process.
2). Polish up the same code - and make it end-user enable-able - so it can be used for the same purpose (finding out if/when a map autosave will happen). This would though require changing each use of the `__func__` macro to something more end-user friendly to identify what is causing the auto-save to be requested.

I guess the simplest approach would be the first one.

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -9461,6 +9470,15 @@ void TLuaInterpreter::pushMapLabelPropertiesToLua(lua_State* L, const TMapLabel&
     lua_pushstring(L, "Pixmap");
     lua_pushstring(L, label.base64EncodePixmap().constData());
     lua_settable(L, -3);
+    lua_pushstring(L, "OnTop");
```
Note that this extends the information returned by `getMapLabel(...)`/`getMapLabels(...)` (to include: whether the map label is on top of everything else; whether it is scaled and whether it is now a temporary one) but it is in a backwards compatible manner... :grinning:

Delwing commented on src/TMap.cpp:
```diff
@@ -1222,32 +1225,35 @@ bool TMap::serialize(QDataStream& ofs, int saveVersion)
-        while (itArea.hasNext()) {
+        while (itArea.hasNext()){
+            // Now we have temporary labels we need to identify areas with
+            // perminent ones:
```
```suggestion
            // permanent ones:
```

Not a blocker of course! :)

SlySven replied:
Actually that spelling mistake appears three times in the code base as I see it now - it is just I do not have the same spell-checking facility in my (currently running OS - Windoze) development environment... *\*sigh\**

Let me throw that fix into another commit.

Delwing reviewed (approved):
I think if we flipped things around... so keep temp labels in separate list, things might be easier, since we just need to call rendering labels twice and don't do other checks, but I might be wrong of course. Besides that I think it's correct.

SlySven committed:
Merge branch 'development' into Add_temporaryMapLabels

SlySven committed:
Revise: correct faulty debugging #if defined(...)

They wouldn't have been showing what I intended should DEBUG_MAPAUTOSAVE be
defined during compilation...!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: fix a spelling mistake I made thrice

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
> I think if we flipped things around... so keep temp labels in separate list, things might be easier, since we just need to call rendering labels twice and don't do other checks, but I might be wrong of course. Besides that I think it's correct.

Keeping them in the same list though keeps the painting code unchanged in `T2DMap` which already takes two passes through the existing lists for the two types of label above and below the rest of the map content.

There is one subtly - I do prevent the map autosave from being activated during the profile loading process so as to prevent some map actions that might otherwise be seen as modifications from being perceived as such during that time. I found this was helpful when the profile loading code presets colours in the map prior to them being replaced by those stored in the map as it is read in.

SlySven committed:
Merge branch 'development' into Add_temporaryMapLabels

SlySven committed:
Revise: fix the third instance of the spelling mistake I made thrice

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

demonnic reviewed (commented):
The temporary map label functionality works, but there seems to be some debugging code left in?

SlySven committed:
Merge branch 'development' into Add_temporaryMapLabels

SlySven committed:
Revise: disable debugging code

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
:books: Wiki documentation updated for:
* **[createMapLabel(...)](https://wiki.mudlet.org/w/Manual:Mapper_Functions#createMapLabel)**
* **[createMapImageLabel(...)](https://wiki.mudlet.org/w/Manual:Mapper_Functions#createMapImageLabel)**
